yabt/buildcontext.py

```diff
@@ -253,6 +255,49 @@ def done_notifier():
 
             return done_notifier
 
+        def make_retry_callback(target: Target):
+            """Return a callable "retry" notifier to
+               report a target as in need of retry.
+               Currently for tests we rebuild the target
+               when it's not necessary."""
+
+            def retry_notifier():
+                """Mark target as retry, re-entering node to end of queue"""
+                if graph_copy.has_node(target.name):
+                    ready_nodes.append(target.name)
+                    produced_event.set()
+
+            return retry_notifier
+
+        def make_fail_callback(target: Target):
+            """Return a callable "fail" notifier to
+               report a target as failed after all retries."""
+
+            def fail_notifier(ex):
+                """Mark target as failed, taking it and decendants
+                   out from the queue"""
+                if graph_copy.has_node(target.name):
+                    self.failed_nodes.append(target.name)
+                    affected_nodes = [target.name]
+                    # removing all predecessors recursively
```

Comment:
no need to manually implement the "recursive get affected nodes" algorithm - just use the built-in one (see https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.ancestors.html).

also see https://github.com/resonai/ybt/blob/master/yabt/graph.py#L340 (but I think you want "ancestors" not "descendants")

Reply:
This creates following error in run-time:
AttributeError: 'DiGraph' object has no attribute 'ancestors'
https://networkx.github.io/documentation/networkx-1.10/reference/classes.digraph.html

Reply:
it's not a method on the DiGraph. it's a function in the dag algorithms module - you pass the graph as an argument